core/src/main/java/cz/xtf/core/git/GitRemoteResolver.java

```diff
@@ -0,0 +1,140 @@
+
+        //get all remote references
+        List<Ref> refs = repository.getRefDatabase().getRefs().stream()
+                .filter(reference -> reference.getName().startsWith("refs/remotes/"))
+                .collect(Collectors.toList());
+
+        List<String> matches = new ArrayList<>();
+        // Walk through all the refs to see if any point to this commit
+        for (Ref ref : refs) {
+            if (ref.getObjectId().equals(commitId)) {
+                matches.add(ref.getName());
+            }
+        }
+
+        if (matches.isEmpty()) {
+            log.debug("No remote references found for the current commit");
+            return;
+        }
+
+        //In case there are multiple matches, we prefer upstream or origin (in this order)
+        List<String> preferredMatches = matches.stream()
+                .filter(reference -> reference.contains("upstream") || reference.contains("origin"))
+                .sorted(Comparator.reverseOrder()) // 1) upstream 2) origin
+                .collect(Collectors.toList());
+
+        if (matches.size() > 1 && !preferredMatches.isEmpty()) {
+            matches = preferredMatches;
+        }
+
+        //branch is string behind the last /
+        reference = matches.stream()
+                .findFirst()
+                .map(ref -> ref.substring(ref.lastIndexOf('/') + 1))
+                .orElse(null);
+
+        log.info("xtf.git.repository.ref got automatically resolved as {}", reference);
+
+        String remote = repository.getRemoteName(matches.get(0));
+        url = getRemoteUrl(repository, remote);
+
+        if (url != null) {
+            log.info("xtf.git.repository.url got automatically resolved as {}", url);
+        }
+
+    }
+
+    /**
+     * given a remote reference, get it's remote URL
+     * 
+     * @param repository git repository
+     * @param remoteReference reference in format "refs/remotes/remote/branch"
+     * @return URL in HTTPS format
+     */
+    private static String getRemoteUrl(Repository repository, String remoteReference) throws URISyntaxException {
+        RemoteConfig remoteConfig = new RemoteConfig(repository.getConfig(), remoteReference);
+        if (remoteConfig.getURIs() == null || remoteConfig.getURIs().isEmpty()) {
+            log.info("Missing URI in git remote ref '{}'", remoteReference);
+            return null;
+        }
+        // we expect a single URI
+        String[] pathTokens = remoteConfig.getURIs().get(0).getPath().split("/");
+        if (pathTokens.length != 2) {
+            log.info("Unexpected path '{}' in URI '{}' of git remote ref '{}'",
+                    remoteConfig.getURIs().get(0).getPath(), remoteConfig.getURIs().get(0), remoteReference);
+            return null;
+        }
+        // the URI must be in HTTPS format
+        return getRepositoryUrl(remoteConfig.getURIs().get(0).getHost(), pathTokens[0], pathTokens[1]);
+    }
+
+    static {
```

Comment:
I'm not in favor to have this executed when it's not needed. Can the whole class be written as a singleton. It would be used like: 
``` 
GitRemoteResolver gitRemoteResolver = GitRemoteResolver.getInstance(); // all initialization happens here
gitRemoteResolver.getRepositoryReference();
gitRemoteResolver.getRepositoryUrl();
```

Reply:
This code is not always executed, it's executed after loading to memory, which according to my testing happens upon first method call of mentioned class

Reply:
Ok, still I think that using Singleton would provide cleaner design.

Reply:
now the methods in Resolver are private, this would just move code you mentioned into `GitRemoteConfig` and would result in no public facing API change

Reply:
There would be no public facing change but internal code could use Singleton, Factory, or Strategy pattern to make it more maintanable. However I'm ok to let it be if we don't anticipate more complex use cases.